### ISSUE TITLE
No support for MCEdit and sponge schematics

### ISSUE THREAD

**Issue by alex9849** [boy0001/FastAsyncWorldedit#1086]
I wrote a plugin that uses WorldEdit schematics to restore WorldGuard regions. It also supports FaWe (API).
I found out, that FaWe can't restore regions with schematics made by worldedit. (Sponge schematic)
It also can't create own schematic files uses the sponge schematic format. (It can create them but not read them)

>[2K
[11:16:38 INFO]: [0;30;1m([0;31;22m[21mFAWE[0;30;1m)[m[0;37;22m Input is corrupt![m
>[2K
[11:16:38 WARN]: java.lang.NullPointerException
>[2K
[11:16:38 WARN]:        at com.boydti.fawe.jnbt.SchematicStreamer.readFully(SchematicStreamer.java:183)
>[2K
[11:16:38 WARN]:        at com.boydti.fawe.jnbt.SchematicStreamer.getClipboard(SchematicStreamer.java:455)
>[2K
[11:16:38 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:69)
>[2K
[11:16:38 WARN]:        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:64)
>[2K
[11:16:38 WARN]:        at my.plugin.adapters.WorldEdit7FaWe.resetBlocks(WorldEdit7FaWe.java:75)
>[2K
[11:16:38 WARN]:        at my.plugin.arm.regions.Region.resetBlocks(Region.java:364)
>[2K
[11:16:38 WARN]:        at my.plugin.arm.regions.Region.resetRegionCommand(Region.java:419)
>[2K
[11:16:38 WARN]:        at my.plugin.arm.Main.onCommand(Main.java:502)
>[2K
[11:16:38 WARN]:        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:44)
>[2K
[11:16:38 WARN]:        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:141)
>[2K
[11:16:38 WARN]:        at org.bukkit.craftbukkit.v1_13_R1.CraftServer.dispatchCommand(CraftServer.java:693)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.PlayerConnection.handleCommand(PlayerConnection.java:1634)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.PlayerConnection.a(PlayerConnection.java:1469)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.PacketPlayInChat.a(PacketPlayInChat.java:45)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.PacketPlayInChat.a(PacketPlayInChat.java:1)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.PlayerConnectionUtils.a(SourceFile:10)
>[2K
[11:16:38 WARN]:        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
>[2K
[11:16:38 WARN]:        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.SystemUtils.a(SourceFile:198)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.MinecraftServer.w(MinecraftServer.java:884)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.DedicatedServer.w(DedicatedServer.java:411)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.MinecraftServer.v(MinecraftServer.java:819)
>[2K
[11:16:38 WARN]:        at net.minecraft.server.v1_13_R1.MinecraftServer.run(MinecraftServer.java:717)
>[2K
[11:16:38 WARN]:        at java.lang.Thread.run(Thread.java:748)
>[2K
[11:16:38 ERROR]: null
org.bukkit.command.CommandException: Unhandled exception executing command 'arm' in plugin AdvancedRegionMarket v1.21
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:46) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:141) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at org.bukkit.craftbukkit.v1_13_R1.CraftServer.dispatchCommand(CraftServer.java:693) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.PlayerConnection.handleCommand(PlayerConnection.java:1634) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.PlayerConnection.a(PlayerConnection.java:1469) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.PacketPlayInChat.a(PacketPlayInChat.java:45) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.PacketPlayInChat.a(PacketPlayInChat.java:1) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.PlayerConnectionUtils.a(SourceFile:10) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_181]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_181]
        at net.minecraft.server.v1_13_R1.SystemUtils.a(SourceFile:198) [minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.MinecraftServer.w(MinecraftServer.java:884) [minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.DedicatedServer.w(DedicatedServer.java:411) [minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.MinecraftServer.v(MinecraftServer.java:819) [minecraft_server.jar:git-Spigot-1503de9-200b239]
        at net.minecraft.server.v1_13_R1.MinecraftServer.run(MinecraftServer.java:717) [minecraft_server.jar:git-Spigot-1503de9-200b239]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
Caused by: java.lang.UnsupportedOperationException: TODO FIXME
        at com.boydti.fawe.jnbt.CorruptSchematicStreamer.recover(CorruptSchematicStreamer.java:100) ~[?:?]
        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:73) ~[?:?]
        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:64) ~[?:?]
        at my.plugin.adapters.WorldEdit7FaWe.resetBlocks(WorldEdit7FaWe.java:75) ~[?:?]
        at my.plugin.arm.regions.Region.resetBlocks(Region.java:364) ~[?:?]
        at my.plugin.arm.regions.Region.resetRegionCommand(Region.java:419) ~[?:?]
        at my.plugin.arm.Main.onCommand(Main.java:502) ~[?:?]
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:44) ~[minecraft_server.jar:git-Spigot-1503de9-200b239]
        ... 15 more
>[2K

**Comment by boy0001** [comment]
If it's a sponge schematic, why is your code trying to use the MCEdit schematic format to load it?

**Comment by alex9849** [comment]
`    public void createSchematic(com.sk89q.worldguard.protection.regions.ProtectedRegion region, String worldname, com.sk89q.worldedit.WorldEdit we) {
        File pluginfolder = Bukkit.getPluginManager().getPlugin("MyFunnyPlugin").getDataFolder();
        File schematicdic = new File(pluginfolder + "/schematics/" + worldname + "/" + region.getId() + ".schematic");
        File schematicfolder = new File(pluginfolder + "/schematics/" + worldname);
        if(schematicdic.exists()){
            schematicdic.delete();
        }

        schematicfolder.mkdirs();

        com.sk89q.worldedit.world.World world = new BukkitWorld(Bukkit.getWorld(worldname));
        CuboidRegion reg = new CuboidRegion(world, region.getMinimumPoint(), region.getMaximumPoint());
        Schematic schem = new Schematic(reg);
        try {
            schematicdic.createNewFile();
            schem.save(schematicdic, ClipboardFormat.SPONGE_SCHEMATIC);
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void resetBlocks(com.sk89q.worldguard.protection.regions.ProtectedRegion region, String worldname, com.sk89q.worldedit.WorldEdit we) {
        File pluginfolder = Bukkit.getPluginManager().getPlugin("MyFunnyPlugin").getDataFolder();
        File file = new File(pluginfolder + "/schematics/" + worldname + "/" + region.getId() + ".schematic");
        World weWorld = new BukkitWorld(Bukkit.getWorld(worldname));
        try {
            ClipboardFormats.findByFile(file).load(file).paste(weWorld, region.getMinimumPoint());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
`

This is the code that should theoretically reset the regions and create the schematics. The plugin should create Sponge schematics and should be able to read every schematic-type that Worldedit and/or FaWe support.

**Comment by itsTurnip** [comment]
I faced same issue today trying to paste WorldEdit schematic to the world.

```
An unexpected error occurred while handling a FAWE command
java.lang.UnsupportedOperationException: TODO FIXME
        at com.boydti.fawe.jnbt.CorruptSchematicStreamer.recover(CorruptSchematicStreamer.java:101) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extent.clipboard.io.SchematicReader.read(SchematicReader.java:74) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extent.clipboard.io.ClipboardFormat.hold(ClipboardFormat.java:128) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.command.SchematicCommands.load(SchematicCommands.java:264) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_211]
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_211]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_211]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_211]
        at com.sk89q.worldedit.util.command.parametric.ParametricCallable.call(ParametricCallable.java:266) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.util.command.SimpleDispatcher.call(SimpleDispatcher.java:134) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.util.command.SimpleDispatcher.call(SimpleDispatcher.java:134) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extension.platform.CommandManager.lambda$handleCommandOnCurrentThread$0(CommandManager.java:422) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extension.platform.CommandManager.handleCommandTask(CommandManager.java:444) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extension.platform.CommandManager.handleCommandOnCurrentThread(CommandManager.java:424) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.sk89q.worldedit.extension.platform.CommandManager.lambda$null$1(CommandManager.java:537) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.boydti.fawe.object.FawePlayer.lambda$runAction$4(FawePlayer.java:299) ~[FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.boydti.fawe.object.task.SimpleAsyncNotifyQueue.operate(SimpleAsyncNotifyQueue.java:25) [FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at com.boydti.fawe.object.task.AsyncNotifyQueue$1.run(AsyncNotifyQueue.java:15) [FastAsyncWorldEdit-bukkit-1.13.138.jar:?]
        at org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftTask.run(CraftTask.java:81) [spigot.jar:git-Spigot-1a3504a-84f3da3]
        at org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52) [spigot.jar:git-Spigot-1a3504a-84f3da3]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_211]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_211]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_211]
```

**Comment by N0tMyFaultOG** [comment]
Your ticket has been closed because it falls into one of the following categories:

- Your issue has been fixed in a newer version of FastAsyncWorldEdit for versions 1.8 - 1.12.2, get the latest build [here](https://ci.athion.net/job/FastAsyncWorldEdit/lastSuccessfulBuild/artifact/target/) matching the platform you need.

- Your issue targets minecraft version 1.13 or higher. This is the wrong repository then. Move to [this repository](https://github.com/IntellectualSites/FastAsyncWorldEdit) and open a bug report here.
BUT BEWARE: You should consider updating to 1.15.2, using the latest Fawe build for 1.15.2 (Build 255 when writing this, get it [here](https://ci.athion.net/job/FastAsyncWorldEdit-1.15/)) and reproducing the issue in a clean environment (fresh setup Spigot or Paper server).

- Your issue has been fixed in a build of FAWE for 1.15 or an issue exists on our new repository.

- You either didn't fill out the template or deleted it.

If none of the bullet points suits, head to the second one.
